### PR TITLE
Feature/implement restaurant order logic 

### PR DESCRIPTION
**Pull request by BartoliniAndBorderCollies**
#### Issue
https://trello.com/c/GaTm1inj

### Description

This branch combines all previous branches in one logic, so that application became one, constant, stable app with all basic functionalities, logic, validations, relations and so on.
From now on `it is possible to make the restaurant order after fulfilling some requirements` (there must be appropriate amount of ingredients on the restaurant stock, menu must be planned and available, if the meal is ordered in restaurant the specific restaurant table must exist and so on).
`Application also counts the total amount to be paid, gives the information what is the order status. Gives the time when food was ordered. To which table it should be delivered, or maybe it is take away order. Besides that personnel know what ingredients are on stock, how many of them. App also updates the amount of ingredients after food is ordered and so on.`

### Motivation

These changes make almost full functionality of the entire app (besides reports which will be done in the next branch) making the application easy to use. Personnel will have nicely presented all necessary information making their work much better. They don't need to remember about anything and they can focus on work. Also application prevents from making mistakes and informs user about the status and overall situation. I believe this will be a great help for the whole restaurant team.

### Changes introduced by this pull request

- refactor of integration and Junit tests taking into account new logic
- refactor of data types in model classes (for example from List(String) ingredients to List(Ingredients) ingredients)
- add new DTOs
- change the relationship between RestaurantOrder and MenuRecord and creation of the third entity: RestaurantOrderMenuRecord and its repository and service
- add custom exception NotEnoughIngredientsException and its handling
- add logic for checking if inventoryItem with specific supplier already exists in db
- `add logic while creating a restaurant order: if menu is provided, if such menu exists in restaurant, are there enough ingredients, checking if provided restaurant table exists, set the order time and order status, update the stock amount, set the total price for the order`
- refactor of findById, findAll, update and delete restaurant order taking into account newly implemented logic

### Testing

- [x] all tests are updated and functional
- [x] coverage for the `whole` application is: 94% classes, 95% lines

### CODE REVIEW

File: src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/InventoryItemController.java [hunk 1]
```diff
@@ -79,7 +80,7 @@ public InventoryItemDTOResponse update(@PathVariable Long id, @RequestBody Inven
      * @throws NotFoundInDatabaseException If the requested item is not found in the database.
      */
     @DeleteMapping("/delete/{id}")
-    public ResponseEntity<?> deleteById(@PathVariable Long id) throws NotFoundInDatabaseException {
+    public ResponseEntity<?> delete(@PathVariable Long id) throws NotFoundInDatabaseException {
```
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/InventoryItemController.java:83]: Was this change required? :)

**Review by JasinskiSz — APPROVED** [review]
There is a lot of changes and I am unable to check everything here. I didn't run the app, but looked through the code and didn't find anything out of ordinary or not working. If you need me to look into something specific here – like looking through a given file, piece of code or something like that, let me know. I might have missed something, as this is a huge PR.
For me this is approvable :) Huge piece of a good job lies there!

**Comment by BartoliniAndBorderCollies** [comment]
great thanks! I would like you to look specifically at RestaurantOrderServiceTest -> JUnit tests

**Comment by JasinskiSz** [comment]
I'll check it out today!

**Comment by JasinskiSz** [comment]
I had to stop in the middle. I'll come back tomorrow with solid results of my investigation :)

File: src/test/java/com/example/Restaurant_Management_System_REST_API/service/RestaurantOrderServiceTest.java [hunk 2]
```diff
@@ -127,39 +145,35 @@ public void update_ShouldThrowNotFoundInDatabaseException_WhenRestaurantOrderWit
...
 
-    //TODO: this below test will need to be updated after merging new branch
     @Test
     public void update_ShouldInteractWithDependenciesCorrectlyAndReturnRestaurantOrderDTO_WhenIdAndRestaurantOrderDTOsGiven()
             throws NotFoundInDatabaseException {
         //Arrange
         Table table = mock(Table.class);
-        MenuRecord mockMenuRecord = mock(MenuRecord.class);
-        List<MenuRecord> menuRecordList = Arrays.asList(mockMenuRecord);
+        String telephoneNumber = "888888888";
```
Comment (JasinskiSz) [inline src/test/java/com/example/Restaurant_Management_System_REST_API/service/RestaurantOrderServiceTest.java]: This is called magic number/string which should be avoided. Instead of this phone number being here it should be contant on top, properly named

File: src/test/java/com/example/Restaurant_Management_System_REST_API/service/RestaurantOrderServiceTest.java [hunk 3]
```diff
@@ -30,55 +31,72 @@ class RestaurantOrderServiceTest {
...
+        // Act and Assert
+        assertThrows(NotFoundInDatabaseException.class, () -> restaurantOrderService.create(restaurantOrderDTO));
+    }
 
-        //Assert
-        assertEquals(restaurantOrderDTO, actual);
+    private RestaurantOrderRequestDTO createRestaurantOrderDTOWithInvalidMeal() {
+        RestaurantOrderRequestDTO restaurantOrderDTO = new RestaurantOrderRequestDTO();
+        restaurantOrderDTO.setMenuRecords(Arrays.asList(new MenuRecordForOrderDTO(1L, "Not valid meal", 1.0)));
+        return restaurantOrderDTO;
```
Comment (JasinskiSz) [inline src/test/java/com/example/Restaurant_Management_System_REST_API/service/RestaurantOrderServiceTest.java]: There is a room to improve by avoiding magic numbers.
An explanation code regarding this matter could look like this:
```
private static final Long INVALID_MEAL_ID = 1L;
private static final String INVALID_MEAL_NAME = "Not valid meal";
private static final Double INVALID_MEAL_PRICE = 1.0;

private RestaurantOrderRequestDTO createRestaurantOrderDTOWithInvalidMeal() {
    RestaurantOrderRequestDTO restaurantOrderDTO = new RestaurantOrderRequestDTO();
    restaurantOrderDTO.setMenuRecords(Arrays.asList(new MenuRecordForOrderDTO(INVALID_MEAL_ID, INVALID_MEAL_NAME, INVALID_MEAL_PRICE)));
    return restaurantOrderDTO;
}
```

**Review by JasinskiSz — COMMENTED** [review]
Looks like I have some tips and found a few places to improve the code in the `RestaurantOrderServiceTest.java`.

1. Tests naming
The test names are quite long and a bit hard to understand for me. These are descriptive – sure! – but are lacking in clean and simple names. 
An example of a better name might be: `delete_ShouldInteractWithDependenciesCorrectlyAndReturnResponseEntity_WhenRestaurantOrderIdIsGiven` -> `delete_ShouldReturnOkResponse_WhenOrderExists`
"ShouldInteractWithDependenciesCorrectly" doesn't mean a lot to me and it's heavy in reading if you know what I mean :)
2. Atomization
When `assertEquals(expected, actual);` is used on type `ResponseEntity` it's checked if whole object is correctly set. This could be improved by checking object's properties – in this case it's body and status. This way it'd be easier to catch what exactly is wrong with the test – is it a status code? Is it a message in body?
3. Minor misspells
There are few misspells I've found. I.e. `nonExistedId` should be `nonExistentId`. Also in this example it might be good idea to specify to which object id belongs. What I mean is to use naming like `nonExistentOrderId`.
4. Avoiding magic numbers
I've explained in one of the comments. [Wikipedia also nicely explains this](https://en.wikipedia.org/wiki/Magic_number_(programming)) stating pros and cons.

**Comment by BartoliniAndBorderCollies** [comment]
"_An example of a better name might be: delete_ShouldInteractWithDependenciesCorrectlyAndReturnResponseEntity_WhenRestaurantOrderIdIsGiven -> delete_ShouldReturnOkResponse_WhenOrderExists_"
aaa okay! I thought that name should exactly indicate what is going on inside the method. The one which you provided: delete_ShouldReturnOkResponse_WhenOrderExists sounds good to me so I can rename those in the new for me manner!

**Comment by BartoliniAndBorderCollies** [comment]
- [x] tests names ready for re-review

**Comment by BartoliniAndBorderCollies** [comment]
- [x] point 2 clear and understood! done and ready for re-review!

**Comment by BartoliniAndBorderCollies** [comment]
point 3 and 4 I will finish tomorrow!

**Comment by BartoliniAndBorderCollies** [comment]
okay point number 3 and 4 also understandable and corrected! ready for re-review!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/InventoryItemController.java (hunk 1 above)
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/InventoryItemController.java:83]: nope but it hurts my eyes ;)

**Review by JasinskiSz — APPROVED** [review]
The changes are looking good. You’re ready to merge!
